### PR TITLE
Use the orchestrator client for ECS and IMDS credentials in aws-config

### PR DESCRIPTION
**Pull request by jdisanti**
This ports the direct uses of the `aws_smithy_client::Client` in aws_config to the orchestrator.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsPresigningDecorator.kt [hunk 1]
```diff
@@ -224,7 +224,10 @@ class AwsPresignedFluentBuilderMethod(
             },
             "alternate_presigning_serializer_registration" to writable {
                 if (presignableOp.hasModelTransforms()) {
-                    rust(".with_operation_plugin(AlternatePresigningSerializerRuntimePlugin)")
```
Comment (rcoh) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsPresigningDecorator.kt:227]: is this related to this PR?

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/HttpChecksumRequiredGenerator.kt [hunk 2]
```diff
@@ -38,7 +39,7 @@ class HttpChecksumRequiredGenerator(
             is OperationSection.AdditionalRuntimePlugins -> writable {
                 section.addOperationRuntimePlugin(this) {
                     rustTemplate(
-                        "#{HttpChecksumRequiredRuntimePlugin}::new()",
```
Comment (rcoh) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/HttpChecksumRequiredGenerator.kt:41]: would you ever not want a shared runtime plugin? Should we add an impl Into or something to make this (and other additional plugin code) cleaner?

we already have `addOperationRuntimePlugin` so you could put the wrapping there

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/IdempotencyTokenGenerator.kt [hunk 3]
```diff
@@ -58,12 +60,14 @@ class IdempotencyTokenGenerator(
                         // then we'll generate one and set it.
                         rustTemplate(
                             """
-                            #{IdempotencyTokenRuntimePlugin}::new(|token_provider, input| {
```
Comment (rcoh) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/IdempotencyTokenGenerator.kt:61]: (same)

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs [hunk 4]
```diff
@@ -235,21 +235,13 @@ impl RuntimePlugins {
...
-    pub fn with_operation_plugin(mut self, plugin: impl RuntimePlugin + 'static) -> Self {
-        insert_plugin!(
-            self.operation_plugins,
-            plugin,
-            SharedRuntimePlugin::new(plugin)
-        );
+    pub fn with_operation_plugin(mut self, plugin: SharedRuntimePlugin) -> Self {
+        insert_plugin!(self.operation_plugins, plugin);
         self
     }
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs:254]: seems like this makes the API a bit more annoying in a bunch of places—I don't necessarily mind making the API tighter here but maybe we could add an extension trait or something?

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs [hunk 5]
```diff
@@ -235,21 +235,13 @@ impl RuntimePlugins {
         Default::default()
     }
 
-    pub fn with_client_plugin(mut self, plugin: impl RuntimePlugin + 'static) -> Self {
-        insert_plugin!(
-            self.client_plugins,
-            plugin,
-            SharedRuntimePlugin::new(plugin)
-        );
+    pub fn with_client_plugin(mut self, plugin: SharedRuntimePlugin) -> Self {
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs:238]: if we end up keeping this change we might want to update the changelog. I think there is already some customer code using this.

**Comment by rcoh** [comment]
the direct changes look good. My main questions / feedback are around the explicit `SharedRuntimePlugin` API change.

File: aws/rust-runtime/aws-config/Cargo.toml [hunk 6]
```diff
@@ -9,7 +9,7 @@ license = "Apache-2.0"
 repository = "https://github.com/awslabs/smithy-rs"
 
 [features]
-client-hyper = ["aws-smithy-client/client-hyper"]
+client-hyper = ["aws-smithy-client/client-hyper", "aws-smithy-runtime/connector-hyper"]
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/Cargo.toml:12]: The connector type on Config is still from aws-smithy-client, so it has to stay here for now. My next PR will remove it since it is cutting over to the newer traits for config.

File: aws/rust-runtime/aws-config/Cargo.toml [hunk 7]
```diff
@@ -27,7 +27,10 @@ aws-smithy-client = { path = "../../sdk/build/aws-sdk/sdk/aws-smithy-client", de
 aws-smithy-http = { path = "../../sdk/build/aws-sdk/sdk/aws-smithy-http" }
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/Cargo.toml:27]: Same answer as above.

File: aws/rust-runtime/aws-config/src/imds/client/error.rs [hunk 8]
```diff
@@ -8,13 +8,14 @@
...
 use aws_smithy_http::body::SdkBody;
 use aws_smithy_http::endpoint::error::InvalidEndpointError;
+use aws_smithy_runtime_api::client::orchestrator::HttpResponse;
 use std::error::Error;
 use std::fmt;
 
 /// Error context for [`ImdsError::FailedToLoadToken`]
 #[derive(Debug)]
 pub struct FailedToLoadToken {
-    source: SdkError<TokenError>,
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client/error.rs:17]: Yeah, I need to update the default generic on that.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/HttpChecksumRequiredGenerator.kt (hunk 2 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/HttpChecksumRequiredGenerator.kt:41]: My thought on this was having the flexibility to actually share runtime plugins. Making it wrap them would make this impossible without a later refactor.

File: aws/rust-runtime/aws-config/src/imds/client/token.rs [hunk 9]
```diff
@@ -47,158 +51,223 @@ const TOKEN_REFRESH_BUFFER: Duration = Duration::from_secs(120);
 
 const X_AWS_EC2_METADATA_TOKEN_TTL_SECONDS: &str = "x-aws-ec2-metadata-token-ttl-seconds";
 const X_AWS_EC2_METADATA_TOKEN: &str = "x-aws-ec2-metadata-token";
+const IMDS_TOKEN_AUTH_SCHEME: AuthSchemeId = AuthSchemeId::new(X_AWS_EC2_METADATA_TOKEN);
 
 /// IMDS Token
 #[derive(Clone)]
 struct Token {
     value: HeaderValue,
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client/token.rs:59]: It wasn't being marked as sensitive, so I fixed that. Will keep the Debug impl just as an additional safeguard.

File: aws/rust-runtime/aws-config/src/imds/client/error.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client/error.rs:17]: Going to do this in a later PR. Doing it now requires modifying a lot of code that will be deleted soon.

File: aws/rust-runtime/aws-config/src/imds/client.rs [hunk 10]
```diff
@@ -415,15 +420,8 @@ impl Builder {
...
-    pub(super) fn build_lazy(self) -> LazyClient {
-        LazyClient {
-            client: OnceCell::new(),
-            builder: self,
-        }
-    }
-
     /// Build an IMDSv2 Client
-    pub async fn build(self) -> Result<Client, BuildError> {
+    pub fn build(self) -> Client {
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client.rs:424]: The only check being done before was the eager endpoint resolution. I think it makes it more consistent with our other clients if endpoint resolution fails on request.

File: aws/rust-runtime/aws-config/src/http_credential_provider.rs [hunk 11]
```diff
@@ -104,198 +96,243 @@ impl Builder {
...
+        //   - Socket errors
+        //   - Networking timeouts
+        //   - 5xx errors
+        //   - Non-parseable 200 responses.
+        let retry_classifiers = RetryClassifiers::new()
+            .with_classifier(HttpCredentialRetryClassifier)
+            // Socket errors and network timeouts
+            .with_classifier(SmithyErrorClassifier::<Error>::new())
+            // 5xx errors
+            .with_classifier(HttpStatusCodeClassifier::default());
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/http_credential_provider.rs:110]: The retry classification has tests for each of socket errors, network timeouts, 5xx errors, and non-parseable 200s. I could see some unspecified corner case changing, but the main guarantees are well tested. I rather reuse the socket/network intermittency classifier since that will likely have a lot of bug fixes made to it over time. I doubt the 5xx status classifier will ever change.

File: aws/rust-runtime/aws-config/src/imds/client/token.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client/token.rs:59]: Fixed in fc8bffc7e6b3849a00c52066a7f991b3ef4f82f7

File: aws/rust-runtime/aws-config/src/imds/client/token.rs [hunk 12]
```diff
@@ -47,158 +51,223 @@ const TOKEN_REFRESH_BUFFER: Duration = Duration::from_secs(120);
...
-/// It will attach the token to the incoming request on the `x-aws-ec2-metadata-token` header.
-#[derive(Clone)]
-pub(super) struct TokenMiddleware {
-    client: Arc<aws_smithy_client::Client<DynConnector, MapRequestLayer<UserAgentStage>>>,
-    token_parser: GetTokenResponseHandler,
-    token: ExpiringCache<Token, ImdsError>,
-    time_source: SharedTimeSource,
-    endpoint: Uri,
-    token_ttl: Duration,
+/// This runtime plugin wires up the necessary componetns to load/cache a token
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client/token.rs]: Fixed in fc8bffc7e6b3849a00c52066a7f991b3ef4f82f7

File: aws/rust-runtime/aws-config/src/imds/client.rs [hunk 13]
```diff
@@ -415,15 +420,8 @@ impl Builder {
...
 
-    pub(super) fn build_lazy(self) -> LazyClient {
-        LazyClient {
-            client: OnceCell::new(),
-            builder: self,
-        }
-    }
-
     /// Build an IMDSv2 Client
-    pub async fn build(self) -> Result<Client, BuildError> {
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client.rs:426]: Fixed in 3acb9014ee412f8384344ec8835b3ef37db65f76

File: aws/rust-runtime/aws-config/src/http_credential_provider.rs [hunk 14]
```diff
@@ -8,35 +8,43 @@
...
-use http::{HeaderValue, Response, Uri};
+use http::{HeaderValue, Response};
 use std::time::Duration;
-use tower::layer::util::Identity;
 
 const DEFAULT_READ_TIMEOUT: Duration = Duration::from_secs(5);
 const DEFAULT_CONNECT_TIMEOUT: Duration = Duration::from_secs(2);
 
+#[derive(Debug)]
+struct OpInput {
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/http_credential_provider.rs]: Fixed in eb507a413d34205c1d49fbdbc3b15c40eb09fe55

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsPresigningDecorator.kt (hunk 1 above)
Comment (jdisanti) [reply aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsPresigningDecorator.kt]: Reverted in 01720a953b12fc775e3f0c37d83d6f6887bba4b8

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/HttpChecksumRequiredGenerator.kt (hunk 2 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/HttpChecksumRequiredGenerator.kt]: All made easier in 01720a953b12fc775e3f0c37d83d6f6887bba4b8

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs (hunk 4 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs]: Fixed in 01720a953b12fc775e3f0c37d83d6f6887bba4b8

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs (hunk 5 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs]: Should be fine after 01720a953b12fc775e3f0c37d83d6f6887bba4b8

File: rust-runtime/aws-smithy-runtime-api/src/shared.rs [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+    fn from_unshared(value: Unshared) -> Self;
+}
+
+/// Like the `Into` trait, but for (efficiently) converting into a shared type.
+///
+/// If the type is already a shared type, it won't be nested in another shared type.
+pub trait IntoShared<Shared> {
+    /// Creates a shared type from an unshared type.
+    fn into_shared(self) -> Shared;
+}
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/shared.rs:28]: I _think_ this should be an associated type? It doesn't make sense for one type to be convertible into multiple different shared types right?

File: rust-runtime/aws-smithy-runtime-api/src/shared.rs [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
+/// Like the `From` trait, but for converting to a shared type.
+pub trait FromUnshared<Unshared> {
+    /// Creates a shared type from an unshared type.
+    fn from_unshared(value: Unshared) -> Self;
+}
+
+/// Like the `Into` trait, but for (efficiently) converting into a shared type.
+///
+/// If the type is already a shared type, it won't be nested in another shared type.
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/shared.rs:24]: these docs should probably contain a definition of what we mean by "shared" type

File: rust-runtime/aws-smithy-runtime-api/src/shared.rs [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+///
+/// struct SharedThing(Arc<dyn Thing>);
+/// impl Thing for SharedThing {}
+/// impl SharedThing {
+///     fn new(thing: impl Thing + 'static) -> Self {
+///         Self(Arc::new(thing))
+///     }
+/// }
+/// impl_shared_conversions!(convert SharedThing from Thing using SharedThing::new);
+/// ```
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/shared.rs:81]: thanks for adding an example!

File: rust-runtime/aws-smithy-runtime-api/src/shared.rs [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+            "precondition"
+        );
+
+        let very_shared_thing: SharedThing = shared_thing.into_shared();
+        assert_eq!(
+            "SharedThing(Thingamajig)",
+            format!("{very_shared_thing:?}"),
+            "it should not nest the shared thing in another shared thing"
+        );
+    }
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/shared.rs:136]: should we assert via downcasting that an already shared type is not double boxed?

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs [hunk 19]
```diff
@@ -617,13 +639,13 @@ impl RuntimeComponentsBuilder {
...
         Self::new("aws_smithy_runtime_api::client::runtime_components::RuntimeComponentBuilder::for_tests")
-            .with_auth_scheme(SharedAuthScheme::new(FakeAuthScheme))
-            .with_auth_scheme_option_resolver(Some(SharedAuthSchemeOptionResolver::new(FakeAuthSchemeOptionResolver)))
-            .with_endpoint_resolver(Some(SharedEndpointResolver::new(FakeEndpointResolver)))
-            .with_http_connector(Some(SharedHttpConnector::new(FakeConnector)))
-            .with_identity_resolver(AuthSchemeId::new("fake"), SharedIdentityResolver::new(FakeIdentityResolver))
+            .with_auth_scheme(FakeAuthScheme)
+            .with_auth_scheme_option_resolver(Some(FakeAuthSchemeOptionResolver))
+            .with_endpoint_resolver(Some(FakeEndpointResolver))
+            .with_http_connector(Some(FakeConnector))
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs:645]: should these methods accept `T` instead of `Option<T>`? I might imagine a `set_` that accepted an option 💭

File: rust-runtime/aws-smithy-runtime-api/src/client/interceptors.rs [hunk 20]
```diff
@@ -618,19 +618,202 @@ impl SharedInterceptor {
     }
 }
 
-impl AsRef<dyn Interceptor> for SharedInterceptor {
-    fn as_ref(&self) -> &(dyn Interceptor + 'static) {
-        self.interceptor.as_ref()
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/client/interceptors.rs:623]: any reason to remove this? seems like a reasonable way to handle it

File: rust-runtime/aws-smithy-runtime-api/src/client/auth.rs [hunk 21]
```diff
@@ -120,6 +121,12 @@ impl AuthSchemeOptionResolver for SharedAuthSchemeOptionResolver {
     }
 }
 
+impl_shared_conversions!(
+    convert SharedAuthSchemeOptionResolver
+    from AuthSchemeOptionResolver
+    using SharedAuthSchemeOptionResolver::new
+);
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/client/auth.rs:128]: I'd probably consider adding `:` in the macro to remove any parsing hazard but this seems to be working fine

Comment (rcoh) [reply rust-runtime/aws-smithy-runtime-api/src/client/auth.rs:128]:
```
convert: SharedAuthSchemeOptionResolver
from: Auth....
```

File: aws/rust-runtime/aws-config/src/imds/client/token.rs [hunk 22]
```diff
@@ -47,158 +51,226 @@ const TOKEN_REFRESH_BUFFER: Duration = Duration::from_secs(120);
...
+
+impl TokenRuntimePlugin {
+    pub(super) fn new(
+        common_plugin: SharedRuntimePlugin,
+        time_source: SharedTimeSource,
+        token_ttl: Duration,
+    ) -> Self {
+        Self {
+            components: RuntimeComponentsBuilder::new("TokenRuntimePlugin")
+                .with_auth_scheme(SharedAuthScheme::new(TokenAuthScheme::new()))
```
Comment (rcoh) [inline aws/rust-runtime/aws-config/src/imds/client/token.rs:89]:
```suggestion
                .with_auth_scheme(TokenAuthScheme::new())
```

File: aws/rust-runtime/aws-config/src/imds/client/token.rs [hunk 23]
```diff
@@ -47,158 +51,226 @@ const TOKEN_REFRESH_BUFFER: Duration = Duration::from_secs(120);
...
+                .with_auth_scheme_option_resolver(Some(SharedAuthSchemeOptionResolver::new(
+                    StaticAuthSchemeOptionResolver::new(vec![IMDS_TOKEN_AUTH_SCHEME]),
```
Comment (rcoh) [inline aws/rust-runtime/aws-config/src/imds/client/token.rs:91]:
```suggestion
                .with_auth_scheme_option_resolver(Some(
                    StaticAuthSchemeOptionResolver::new(vec![IMDS_TOKEN_AUTH_SCHEME],
```

**Review by rcoh — APPROVED** [review]
I like IntoShared! We should probably ensure that you get decent type hints / docs in various editors at some point

File: rust-runtime/aws-smithy-runtime-api/src/shared.rs (hunk 15 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/shared.rs:28]: I basically made these traits equivalent to `From`/`Into`. I guess it makes sense to restrict it to one shared type, but typing `IntoShared<Shared = MySharedType>` everywhere kind of sucks. I wish we had trait aliases.

File: rust-runtime/aws-smithy-runtime-api/src/shared.rs (hunk 18 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/shared.rs:136]: This seems to be very non-trivial.

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs (hunk 19 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs:645]: It already has the the `set_` variants. I can't remember why I did it this way, but it looks like nothing is calling them with `None`. So I think this is a sane change to make.

Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs:645]: Thinking this should be done in a follow-up PR.

File: rust-runtime/aws-smithy-runtime-api/src/shared.rs (hunk 16 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/shared.rs:105]: Improved in baad17532dc25a5daf2d0f7121676ff5a85da8cb

File: aws/rust-runtime/aws-config/src/imds/client/token.rs (hunk 22 above)
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client/token.rs]: Fixed in 1d68fdfda697e1e50c83951acfef78bebdcdc6d3

File: aws/rust-runtime/aws-config/src/imds/client/token.rs (hunk 23 above)
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/imds/client/token.rs]: Fixed in 1d68fdfda697e1e50c83951acfef78bebdcdc6d3
